File: year_2023/day_01/day01.py
import regex as re

from tools import loader, parsers


def part_1(data: list[str]) -> int:
    """
    >>> print(part_1(parsers.lines('test.txt')))
    142"""
    result = 0
    for line in data:
        nums = re.findall(r'\d', line)
        result += int(nums[0] + nums[-1])
    return result


def part_2(data: list[str]) -> int:
    """
    >>> print(part_2(parsers.lines('test2.txt')))
    281"""
    words = {'one': '1', 'two': '2', 'three': '3',
             'four': '4', 'five': '5', 'six': '6',
             'seven': '7', 'eight': '8', 'nine': '9'}
    re_words = re.compile('|'.join(words.keys()) + r'|\d')
    result = 0
    for line in data:
        matches = re.findall(re_words, line, overlapped=True)
        nums = [i if i.isdigit() else words[i] for i in matches]
        result += int(nums[0] + nums[-1])
    return result


print(part_1(parsers.lines(loader.get())))  # 55002
print(part_2(parsers.lines(loader.get())))  # 55093

File: year_2023/day_01/day01.go
package main

import (
	"github.com/Nyaaa/advent-of-code/tools"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

func solve(input string, part2 bool) int {
	result := 0
	nums := regexp.MustCompile(`\d`)
	words := map[string]string{"one": "o1e", "two": "t2o", "three": "t3e",
		"four": "f4r", "five": "f5e", "six": "s6x",
		"seven": "s7n", "eight": "e8t", "nine": "n9e"}

	for _, line := range tools.ReadLines(input) {
		if part2 {
			for from, to := range words {
				line = strings.ReplaceAll(line, from, to)
			}
		}

		digits := nums.FindAllString(line, -1)
		value := fmt.Sprint(digits[0], digits[len(digits)-1])
		valueInt, _ := strconv.Atoi(value)
		result += valueInt
	}

	return result
}

func main() {
	data := tools.GetData(2023, 1)

	fmt.Println(solve("test.txt", false), solve("test2.txt", true))
	fmt.Println(solve(data, false), solve(data, true))
}
